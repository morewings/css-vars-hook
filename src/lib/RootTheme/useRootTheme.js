import {useRootContext} from './RootContext';

export const useRootTheme = () => useRootContext();
