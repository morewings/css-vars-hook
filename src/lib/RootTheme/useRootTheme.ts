import {useCallback, useMemo, useRef} from 'react';
import type {CSSProperties} from 'react';
import type {ThemeType} from 'css-vars-hook';

import {
    createStyleObject,
    getRootVariable,
    removeRootVariable,
    setRootVariable,
} from '../utils';
import type {HookInterface} from './HookInterfaceType';
import type {UnitType} from '../UnitType';

/**
 * @private
 * Logic for root theme handling such as updates and CSS style creation
 */
export const useRootTheme = (
    theme: ThemeType
): HookInterface & {style: CSSProperties} => {
    const themeRef = useRef(theme);

    const setTheme = useCallback((nextTheme: ThemeType) => {
        Object.keys(nextTheme).forEach(key => {
            setRootVariable(key, nextTheme[key]);
        });

        themeRef.current = nextTheme;
    }, []);

    const getTheme = useCallback(() => themeRef.current, []);

    const getVariable = useCallback(
        (variableName: string) => getRootVariable(variableName),
        []
    );
    const setVariable = useCallback((variableName: string, value: UnitType) => {
        setRootVariable(variableName, value);
        themeRef.current = {
            ...themeRef.current,
            [variableName]: value,
        };
    }, []);

    const removeVariable = useCallback((variableName: string) => {
        removeRootVariable(variableName);
        const nextTheme = {...themeRef.current};
        delete nextTheme[variableName];
        themeRef.current = nextTheme;
    }, []);

    const style = useMemo(() => createStyleObject(themeRef.current), []);

    return {
        /** Effect to apply new theme to the application */
        setTheme,
        /** CSSProperties object to apply to container div */
        style,
        /** Get current theme */
        getTheme,
        /** Get variable value within active theme */
        getVariable,
        /** Effect to set new variable value within active theme */
        setVariable,
        /** Effect to remove variable within active theme */
        removeVariable,
    };
};
