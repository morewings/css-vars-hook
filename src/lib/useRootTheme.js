import {useInsertionEffect} from 'react';
import {setRootVariable, removeRootVariable} from 'lib/utils';

export const useRootTheme = (theme = {}) => {
  // const root = useRef(document.querySelector(':root'));
  useInsertionEffect(() => {
    Object.keys(theme).forEach(key => {
      setRootVariable(key, theme[key]);
    });
  }, [theme]);
  return {
    setRootVariable,
    removeRootVariable,
  };
};
