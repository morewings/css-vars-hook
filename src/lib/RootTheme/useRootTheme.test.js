import {renderHook} from '@testing-library/react-hooks';
import {useRootTheme} from './useRootTheme';

const theme = {
  foo: 'bar',
};

describe('useRootTheme', () => {
  let root;

  beforeEach(() => {
    // eslint-disable-next-line fp/no-mutation
    root = document.querySelector(':root');
  });

  it('renders and applies theme', () => {
    const {result} = renderHook(() => useRootTheme(theme));
    expect(result.current).toMatchSnapshot();
    expect(window.getComputedStyle(root).cssText).toMatchSnapshot();
  });

  it('`setRootVariable` sets CSS variable value', () => {
    const {result} = renderHook(() => useRootTheme(theme));
    const varName = 'foo';
    const varValue = 'baz';
    result.current.setRootVariable(varName, varValue);
    expect(root.style.getPropertyValue(`--${varName}`)).toBe(varValue);
  });

  it('`removeRootVariable` removes CSS variable', () => {
    const {result} = renderHook(() => useRootTheme(theme));
    const varName = 'foo';
    result.current.removeRootVariable(varName);
    expect(root.style.getPropertyValue(`--${varName}`)).toBe('');
  });
});
