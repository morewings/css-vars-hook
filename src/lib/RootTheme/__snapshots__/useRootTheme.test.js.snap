// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRootTheme renders and applies theme 1`] = `undefined`;

exports[`useRootTheme renders and applies theme 2`] = `"visibility: visible;"`;
